File: bf.js
// Author: Suneel Freimuth
// Run with Node. Tested with Node 10 as the earliest version.

const readline = require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
})

class NoInputProvided extends Error {
    constructor(...args) {
        super(...args)
    }
}

const input = prompt => new Promise(resolve => {
    rl.question(prompt, resolve)
})

async function interpret(code, tape_length=30000) {
    const tape = Array(tape_length).fill(0)
    let pdata = 0
    let pinst = 0
    while (pinst < code.length) {
        const inst = code[pinst]
        switch (inst) {
            case '>':
                pdata++
                break
            case '<':
                if (pdata > 0) {
                    pdata--
                }
                break
            case '+':
                tape[pdata]++
                break
            case '-':
                tape[pdata]--
                break
            case '.':
                process.stdout.write( Buffer.from([tape[pdata]]) )
                break
            case ',':
                const inp = (await input('Input>')).trimEnd()
                if (inp.length > 0) {
                    tape[pdata] = inp.codePointAt(0) & 0xFF // Constrain to 1 byte.
                } else {
                    throw new NoInputProvided()
                }
                break
            case '[':
                if (tape[pdata] === 0) {
                    let bracket_stack = 0
                    while (true) {
                        pinst += 1
                        if (code[pinst] === '[') {
                            bracket_stack++
                        } else if (code[pinst] === ']' && bracket_stack > 0) {
                            bracket_stack--
                        } else if (code[pinst] === ']' && bracket_stack === 0) {
                            pinst++
                            break
                        }
                    }
                }
                break
            case ']':
                if (tape[pdata] !== 0) {
                    let bracket_stack = 0
                    while (true) {
                        pinst--
                        if (code[pinst] === ']')
                            bracket_stack++
                        else if (code[pinst] === '[' && bracket_stack > 0)
                            bracket_stack--
                        else if (code[pinst] === '[' && bracket_stack === 0)
                            break
                    }
                }
                break
        }
        pinst++
    }
}

async function main() {
    if (process.argv.length > 2) {
        await interpret(process.argv[2])
    } else {
        while (true) {
            const inp = await input('>')
            await interpret(inp)
            console.log()
        }
    }
}

main().finally(() => {
    rl.close()
})

File: bf.rs
/* Language: Rust
 * Run with
 *   rustc bf.rs
 *   ./bf
 */

use std::io::Write;
use std::env;

const TAPE_LEN: usize = 30_000;

fn interpret(stdin: &std::io::Stdin, stdout: &mut std::io::Stdout, code: &Vec<u8>) {
    let mut tape: [u8; TAPE_LEN] = [0; TAPE_LEN];
    let mut pdata: usize = 0;
    let mut pinst: usize = 0;
    while pinst < code.len() {
        match code[pinst] as char {
            '>' => {
                if pdata < TAPE_LEN - 1 {
                    pdata += 1;
                }
            },
            '<' => {
                if pdata > 0 {
                    pdata -= 1;
                }
            },
            '+' => {
                if tape[pdata] < 255 {
                    tape[pdata] += 1;
                }
            },
            '-' => {
                if tape[pdata] > 0 {
                    tape[pdata] -= 1;
                }
            },
            '.' => {
                stdout.write(&tape[pdata..pdata+1]).expect("wut");
            },
            ',' => {
                print!("Input>");
                let mut input = String::new();
                stdin.read_line(&mut input).expect("wut");
                tape[pdata] = input.as_bytes()[0];
            },
            '[' => {
                if tape[pdata] == 0 {
                    let mut bracket_stack: u32 = 0;
                    while pinst < code.len() {
                        pinst += 1;
                        if code[pinst] as char == '[' {
                            bracket_stack += 1;
                        } else if code[pinst] as char == ']' && bracket_stack > 0 {
                            bracket_stack -= 1;
                        } else if code[pinst] as char == ']' && bracket_stack == 0 {
                            pinst += 1;
                            break;
                        }
                    }
                }
            },
            ']' => {
                if tape[pdata] != 0 {
                    let mut bracket_stack: u32 = 0;
                    while pinst > 0 {
                        pinst -= 1;
                        if code[pinst] as char == ']' {
                            bracket_stack += 1;
                        } else if code[pinst] as char == '[' && bracket_stack > 0 {
                            bracket_stack -= 1;
                        } else if code[pinst] as char == '[' && bracket_stack == 0 {
                            break;
                        }
                    }
                }
            },
            _ => {}
        }
        pinst += 1;
    }
    stdout.write(b"\n").expect("wut");
}


fn main() {
    let args: Vec<String> = env::args().collect();
    let stdin = std::io::stdin();
    let mut stdout = std::io::stdout();
    if args.len() > 1 {
        interpret(&stdin, &mut stdout, &args[1].clone().into_bytes());
    } else {
        loop {
            stdout.write(b">").expect("wut");
            stdout.flush().expect("wut");
            let mut input = String::new();
            stdin.read_line(&mut input).expect("wut");
            interpret(&stdin, &mut stdout, &input.into_bytes());
        }
    }
}
